File: libraries/python/ifcx/converters/dxf_tokenizer.py
"""DXF ASCII tokenizer -- pure Python, no external dependencies.

Reads a DXF text (ASCII) file and yields ``(group_code, typed_value)`` pairs.
"""

from __future__ import annotations

from typing import Generator, Union

# Type alias for a parsed DXF value.
DxfValue = Union[str, int, float, bool]


def _value_type_for_code(code: int) -> str:
    """Return the expected Python type key for a DXF group code.

    Returns one of: 'str', 'float', 'int', 'bool'.
    """
    if 0 <= code <= 9:
        return "str"
    if 10 <= code <= 39:
        return "float"
    if 40 <= code <= 59:
        return "float"
    if 60 <= code <= 79:
        return "int"
    if 90 <= code <= 99:
        return "int"
    if code == 100:
        return "str"
    if code == 102:
        return "str"
    if code == 105:
        return "str"
    if 110 <= code <= 149:
        return "float"
    if 160 <= code <= 169:
        return "int"
    if 170 <= code <= 179:
        return "int"
    if 210 <= code <= 239:
        return "float"
    if 270 <= code <= 289:
        return "int"
    if 290 <= code <= 299:
        return "bool"
    if 300 <= code <= 309:
        return "str"
    if 310 <= code <= 319:
        return "str"
    if 320 <= code <= 369:
        return "str"
    if 370 <= code <= 379:
        return "int"
    if 380 <= code <= 389:
        return "int"
    if 390 <= code <= 399:
        return "str"
    if 410 <= code <= 419:
        return "str"
    if 420 <= code <= 429:
        return "int"
    if 430 <= code <= 439:
        return "str"
    if 440 <= code <= 449:
        return "int"
    if code == 999:
        return "str"
    if 1000 <= code <= 1009:
        return "str"
    if 1010 <= code <= 1059:
        return "float"
    if 1060 <= code <= 1071:
        return "int"
    # Default: treat as string
    return "str"


def _cast_value(code: int, raw: str) -> DxfValue:
    """Convert a raw string value to the appropriate Python type."""
    vtype = _value_type_for_code(code)
    if vtype == "float":
        try:
            return float(raw)
        except ValueError:
            return 0.0
    if vtype == "int":
        try:
            return int(raw)
        except ValueError:
            return 0
    if vtype == "bool":
        try:
            return bool(int(raw))
        except ValueError:
            return False
    # string
    return raw


def tokenize(content: str) -> Generator[tuple[int, DxfValue], None, None]:
    """Yield ``(group_code, typed_value)`` pairs from DXF ASCII text.

    Handles both DOS (CRLF) and Unix (LF) line endings.
    """
    # Normalize line endings to LF, then split.
    lines = content.replace("\r\n", "\n").replace("\r", "\n").split("\n")
    idx = 0
    length = len(lines)

    while idx + 1 < length:
        code_str = lines[idx].strip()
        val_str = lines[idx + 1].strip()
        idx += 2

        if code_str == "":
            continue

        try:
            code = int(code_str)
        except ValueError:
            continue

        yield code, _cast_value(code, val_str)

File: libraries/csharp/Converters/Dxf/DxfTokenizer.cs
namespace Ifcx.Converters.Dxf;

/// <summary>
/// A single DXF group code / value pair.
/// </summary>
public readonly record struct DxfToken(int Code, object Value)
{
    public string StringValue => Value is string s ? s : Value?.ToString() ?? "";
    public double DoubleValue => Value is double d ? d : Convert.ToDouble(Value);
    public int IntValue => Value is int i ? i : Convert.ToInt32(Value);
    public bool BoolValue => Value is bool b ? b : Convert.ToBoolean(Value);
}

/// <summary>
/// Pure C# DXF ASCII tokenizer. Yields (groupCode, typedValue) pairs.
/// No external dependencies.
/// </summary>
public static class DxfTokenizer
{
    /// <summary>
    /// Determines the expected value type for a DXF group code.
    /// Returns "str", "float", "int", or "bool".
    /// </summary>
    public static string ValueTypeForCode(int code) => code switch
    {
        >= 0 and <= 9 => "str",
        >= 10 and <= 39 => "float",
        >= 40 and <= 59 => "float",
        >= 60 and <= 79 => "int",
        >= 90 and <= 99 => "int",
        100 => "str",
        102 => "str",
        105 => "str",
        >= 110 and <= 149 => "float",
        >= 160 and <= 169 => "int",
        >= 170 and <= 179 => "int",
        >= 210 and <= 239 => "float",
        >= 270 and <= 289 => "int",
        >= 290 and <= 299 => "bool",
        >= 300 and <= 309 => "str",
        >= 310 and <= 319 => "str",
        >= 320 and <= 369 => "str",
        >= 370 and <= 379 => "int",
        >= 380 and <= 389 => "int",
        >= 390 and <= 399 => "str",
        >= 410 and <= 419 => "str",
        >= 420 and <= 429 => "int",
        >= 430 and <= 439 => "str",
        >= 440 and <= 449 => "int",
        999 => "str",
        >= 1000 and <= 1009 => "str",
        >= 1010 and <= 1059 => "float",
        >= 1060 and <= 1071 => "int",
        _ => "str",
    };

    /// <summary>
    /// Cast a raw string value to the appropriate type based on group code.
    /// </summary>
    public static object CastValue(int code, string raw)
    {
        var vtype = ValueTypeForCode(code);
        return vtype switch
        {
            "float" => double.TryParse(raw, System.Globalization.NumberStyles.Float,
                System.Globalization.CultureInfo.InvariantCulture, out var d) ? d : 0.0,
            "int" => int.TryParse(raw, out var i) ? i : 0,
            "bool" => int.TryParse(raw, out var b) && b != 0,
            _ => raw,
        };
    }

    /// <summary>
    /// Tokenize DXF ASCII text content into group code / value pairs.
    /// Handles both CRLF and LF line endings.
    /// </summary>
    public static IEnumerable<DxfToken> Tokenize(string content)
    {
        // Normalize line endings
        var text = content.Replace("\r\n", "\n").Replace("\r", "\n");
        var lines = text.Split('\n');
        var idx = 0;
        var length = lines.Length;

        while (idx + 1 < length)
        {
            var codeStr = lines[idx].Trim();
            var valStr = lines[idx + 1].Trim();
            idx += 2;

            if (string.IsNullOrEmpty(codeStr))
                continue;

            if (!int.TryParse(codeStr, out var code))
                continue;

            yield return new DxfToken(code, CastValue(code, valStr));
        }
    }
}
